cf-execd/cf-execd-runagent.c

```diff
@@ -0,0 +1,150 @@
+  (COSL) may apply to this file if you as a licensee so wish it. See
+  included file COSL.txt.
+*/
+
+#include <stdio.h>
+
+#include <logging.h>
+#include <buffer.h>
+#include <json.h>
+#include <pipes.h>
+#include <alloc.h>
+#include <file_lib.h>
+#include <known_dirs.h>
+
+#include <cf-execd-runagent.h>
+
+
+/**
+ * Handle request to run cf-runagent.
+ *
+ * @note Expected to be called from a forked child process (blocks and forks).
+ */
+void HandleRunagentRequest(int conn_fd)
+{
+    FILE *conn_file = fdopen(conn_fd, "r+");
+    if (conn_file == NULL)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to create file stream from the connection file descriptor: %s",
+            GetErrorStr());
+
+        /* SIGH: There's no dprintf() on exotics. */
+        char error_buf[256];
+        int length = snprintf(error_buf, sizeof(error_buf),
+                              "{\"error\": \"Failed to create file stream from the connection file descriptor: %s\"}",
+                              GetErrorStr());
+        assert(length < sizeof(error_buf));
+        write(conn_fd, error_buf, MIN(sizeof(error_buf), (size_t) length));
+        close(conn_fd);
+        return;
+    }
+
+    char *request = NULL;
+    size_t line_size = 0;
+    ssize_t n_read = getline(&request, &line_size, conn_file);
+    if (n_read <= 0)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to read the runagent request: %s",
+            (n_read < 0) ? GetErrorStr() : "no data");
+        fprintf(conn_file, "{\"error\": \"Failed to read the runagent request: %s\"}",
+                (n_read < 0) ? GetErrorStr() : "no data");
+        free(request);
+        fclose(conn_file);
+        return;
+    }
+
+    /* TODO: '--raw' for just getting output from cf-runagent directly? */
+    for (char *c = (request + n_read - 1); (c > request) && (*c != '\n'); c--)
+    {
+        if (*c == '\n')
+        {
+            *c = '\0';
+        }
+    }
+
+    char *command;
+    int ret = xasprintf(&command, "%s%ccf-runagent -b -H %s", GetBinDir(), FILE_SEPARATOR, request);
+    assert(ret != -1); /* should never happen */
+    if (ret == -1)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to construct cf-runagent command for '%s'", request);
+        fprintf(conn_file, "{\"error\": \"Failed to construct cf-runagent command for '%s'\"}", request);
+        free(request);
+        fclose(conn_file);
+        return;
+    }
+    free(request);
+
+    FILE *runagent_output = cf_popen(command, "r", true);
+    free(command);
+
+    if (runagent_output == NULL)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to start cf-runagent and connect to its output");
+        fprintf(conn_file, "{\"error\": \"Failed to start cf-runagent and connect to its output\"}");
+        fclose(conn_file);
+        return;
+    }
+
+    Buffer *collected_output = BufferNewWithCapacity(1024);
+    char output_buffer[CF_BUFSIZE];
+    size_t read_bytes = fread(output_buffer, 1, sizeof(output_buffer), runagent_output);
```

Comment:
Personally I would use `sizeof(char)` to avoid using magic numbers although it's pretty obvios what the `1` means in this case.

Reply:
I'd have to call the variable `read_chars` then. :wink: